### PR TITLE
Add Cutlass NVFP4 Gemm to `nvfuser_direct` python bindings

### PR DESCRIPTION
**Pull request by rdspring1**
This PR adds `cutlass_nvfp4_scaled_mm` to the `nvfuser_direct` python bindings, which support nvfp4 gemm.

PR Stack:
- #4722
- #4676 **<<< This PR.**
- #4662

### CODE REVIEW

File: cutlass/nvfp4_quantize.cu [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+
+// Fast reciprocal.
+inline __device__ float reciprocal_approximate_ftz(float a) {
+  float b;
+  asm volatile("rcp.approx.ftz.f32 %0, %1;\n" : "=f"(b) : "f"(a));
+  return b;
+}
+
+template <class SFType, int CVT_FP4_NUM_THREADS_PER_SF>
+__device__ uint8_t* cvt_quant_to_fp4_get_sf_out_offset(
```
Comment (crcrpar) [inline cutlass/nvfp4_quantize.cu]: Is this based on https://github.com/NVIDIA/TensorRT-LLM/blob/d135f5993dd12ad25ecd5c58a8e7a0e3d05bfc4e/cpp/tensorrt_llm/kernels/quantization.cuh#L617? I'm just curious

File: python/nvfuser_direct/cutlass.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    device = input.device
+
+    assert n % block_size == 0, f"last dim has to be multiple of 16, but got {n}."
+    assert input.dtype in (
+        torch.float16,
+        torch.bfloat16,
+    ), f"input.dtype needs to be fp16 or bf16 but got {input.dtype}."
+
+    # Two fp4 values will be packed into an uint8.
+    output = torch.empty((m, n // 2), device=device, dtype=torch.uint8)
```
Comment (crcrpar) [inline python/nvfuser_direct/cutlass.py]: now PyTorch has `torch.float4_e2m1fn_x2` dtype so maybe in future we want to use it instead of `torch.uint8`

**Comment by rdspring1** [comment]
!test

File: cutlass/nvfp4_quantize.cu (hunk 1 above)
Comment (rdspring1) [reply cutlass/nvfp4_quantize.cu]: It is from https://github.com/sgl-project/sglang/pull/3899/files#diff-cf3ea66f6471a215d708c9ce6facfe7d3678de48d21277ced510467bfeb80df3R128.

Comment (crcrpar) [reply cutlass/nvfp4_quantize.cu]: understood. shouldn't we include the license they use in this file?

File: cutlass/nvfp4_quantize.cu [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+    torch::Tensor& output_scale,
+    const torch::Tensor& input,
+    const torch::Tensor& input_scale) {
+  int32_t m = (int32_t)input.size(0);
+  int32_t n = (int32_t)input.size(1);
+
+  NVF_CHECK(n % 16 == 0, "The N dimension must be multiple of 16.");
+
+  int multiProcessorCount = getMultiProcessorCount();
+
```
Comment (xwang233) [inline cutlass/nvfp4_quantize.cu]:
```suggestion
  // Assuming NVF_CHECK is a macro to check conditions
  #define NVF_CHECK(condition, message) \
    do { if (!(condition)) throw std::runtime_error(message); } while (0)

```
**[Suggested by AI]** The code introduces a macro named `NVF_CHECK`. This macro takes a boolean `condition` and a `message` string as input.  If the `condition` is false, it throws a `std::runtime_error` exception with the provided `message`.  This is a common pattern for adding runtime assertions or checks to code, likely to catch unexpected states during execution and provide informative error messages.

Essentially, this adds a way to proactively halt execution and report errors if certain conditions aren't met, which likely resolved a build error caused by an unhandled or unexpected state.

File: cutlass/nvfp4_quantize.cu (hunk 1 above)
Comment (rdspring1) [reply cutlass/nvfp4_quantize.cu]: I'm not sure if we're going to use this kernel if we can fuse it nvfuser.

Comment (rdspring1) [reply cutlass/nvfp4_quantize.cu]: Do we need a copyright for the cutlass kernel or should we repurpose cutlass example 79b for the nvfp4 gemm?

File: cutlass/nvfp4_quantize.cu (hunk 3 above)
Comment (jjsjann123) [reply cutlass/nvfp4_quantize.cu]: I don't understand what the agent is suggesting....

Comment (crcrpar) [reply cutlass/nvfp4_quantize.cu]: we do have `NVF_CHECK` defined in https://github.com/NVIDIA/Fuser/blob/1a902f7dbcf7a401282f8218b01311bc5eddd794/csrc/exceptions.h#L285-L292 

false alarm?

Comment (crcrpar) [reply cutlass/nvfp4_quantize.cu]: ah I think I got what the suggestion means. This file doesn't include exceptions.h

File: cutlass/nvfp4_quantize.cu [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+void nvfp4_quantize(
+    torch::Tensor& output,
+    torch::Tensor& output_scale,
+    const torch::Tensor& input,
+    const torch::Tensor& input_scale) {
+  int32_t m = (int32_t)input.size(0);
+  int32_t n = (int32_t)input.size(1);
+
+  NVF_CHECK(n % 16 == 0, "The N dimension must be multiple of 16.");
```
Comment (xwang233) [inline cutlass/nvfp4_quantize.cu]:
```suggestion
  TORCH_CHECK(n % 16 == 0, "The N dimension must be multiple of 16.");
```
**[Suggested by AI]** The code change replaces `NVF_CHECK` with `TORCH_CHECK`. Both appear to be assertion macros, likely used for error checking. The change likely standardizes the error checking mechanism to use the `TORCH_CHECK` macro consistently within the PyTorch framework, potentially for better integration with PyTorch's error handling and reporting.  The core functionality (checking if `n` is a multiple of 16) remains the same.

File: cutlass/nvfp4_quantize.cu (hunk 3 above)
Comment (crcrpar) [reply cutlass/nvfp4_quantize.cu]: @rdspring1 could you fix this? I locally failed to build this branch due to the missing include of `exceptions.h`

File: tests/python_direct/test_cutlass_nvfp4_gemm.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    )
+    return torch.matmul(a_in_dtype, b_in_dtype.t())
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("shape", SHAPES)
+@torch.inference_mode()
+def test_nvfp4_gemm(
+    dtype: torch.dtype,
+    shape: tuple[int, int],
```
Comment (crcrpar) [inline tests/python_direct/test_cutlass_nvfp4_gemm.py]: nit-picking

```suggestion
    shape: tuple[int, int, int],
```

**Comment by rdspring1** [comment]
!test

File: python/nvfuser/__init__.py [hunk 6]
```diff
@@ -3,10 +3,13 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
 import sys
+import warnings
 
-assert (
-    "nvfuser_direct" not in sys.modules
-), "Cannot import nvfuser if nvfuser_direct module is already imported."
+if "nvfuser_direct" in sys.modules:
```
Comment (rdspring1) [inline python/nvfuser/__init__.py]: @crcrpar You can import `nvfuser` and `nvfuser_direct` together.

File: python/nvfuser_direct/cutlass.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+def cutlass_nvfp4_scaled_mm(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    block_scale_a: torch.Tensor,
+    block_scale_b: torch.Tensor,
+    alpha: torch.Tensor,
+    out_dtype: torch.dtype,
+) -> torch.Tensor:
+    assert a.ndim == 2 and b.ndim == 2
+    m, n = a.shape[0], b.shape[0]
```
Comment (IvanYashchuk) [inline python/nvfuser_direct/cutlass.py]: Usually, "mm" operation takes [m, k] x [k, n] -> [m, n]. What seems to be implemented here is "mm_NT" with `b` interpreted to be transposed during matrix multiplication. A more accurate name for PyTorch users would be "scaled_linear".

What's the expected layout of `block_scale_b`? Should it be the scaling factors for the transposed matrix `b`?

File: tests/python_direct/test_cutlass_nvfp4_quantize.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+from nvfuser_direct import cutlass_nvfp4_quantize
+
+if torch.cuda.get_device_capability() < (10, 0):
+    pytest.skip(
+        reason="Nvfp4 Requires compute capability of 10 or above.",
+        allow_module_level=True,
+    )
+
+DTYPES = [torch.float16, torch.bfloat16]
+SHAPES = [(128, 64), (128, 128), (256, 64), (256, 128)]
```
Comment (crcrpar) [inline tests/python_direct/test_cutlass_nvfp4_quantize.py]: As this only covers 2D shapes, how about checking that input tensors are 2D?

**Comment by rdspring1** [comment]
!test

File: cutlass/nvf_cutlass.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace nvfuser::cutlass_kernels {
+
+void nvfp4_scaled_mm(
+    torch::Tensor& output,
+    const torch::Tensor& a,
+    const torch::Tensor& b,
+    const torch::Tensor& scales_a,
+    const torch::Tensor& scales_b,
+    const torch::Tensor& alpha);
```
Comment (jjsjann123) [inline cutlass/nvf_cutlass.h]: I thought the cutlass kernel supported nvfp4 outputs?

File: python/nvfuser_direct/cutlass.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    out = torch.empty((m, n), dtype=out_dtype, device=a.device)
```
Comment (jjsjann123) [inline python/nvfuser_direct/cutlass.py]: nitpick: would appreciate to have the output allocation part done inside the kernel.

File: cutlass/nvfp4_scaled_mm.cu [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+constexpr auto SF_DTYPE = at::ScalarType::Float8_e4m3fn;
+
+void nvfp4_scaled_mm(
+    torch::Tensor& output,
+    torch::Tensor const& a,
+    torch::Tensor const& b,
+    torch::Tensor const& scales_a,
+    torch::Tensor const& scales_b,
+    torch::Tensor const& alpha) {
+  CHECK_INPUT(a, FLOAT4_E2M1X2, "a");
```
Comment (jjsjann123) [inline cutlass/nvfp4_scaled_mm.cu]: Would be great if we expose this as a `canExecute` function.

File: CMakeLists.txt [hunk 12]
```diff
@@ -656,19 +728,28 @@ if(BUILD_PYTHON)
     CXX_STANDARD_REQUIRED ON
     CXX_VISIBILITY_PRESET hidden
     INSTALL_RPATH
-    "$ORIGIN/lib:$ORIGIN/../nvidia/cuda_runtime/lib:$ORIGIN/../nvidia/cuda_nvrtc/lib:$ORIGIN/../../nvidia/cuda_cupti/lib:$ORIGIN/../torch/lib"
+    "$ORIGIN/lib:$ORIGIN/../nvfuser_common/lib:$ORIGIN/../nvidia/cuda_runtime/lib:$ORIGIN/../nvidia/cuda_nvrtc/lib:$ORIGIN/../../nvidia/cuda_cupti/lib:$ORIGIN/../torch/lib"
```
Comment (jjsjann123) [inline CMakeLists.txt:731]: sneaky... I didn't see this one earlier. :face_holding_back_tears:

Comment (rdspring1) [reply CMakeLists.txt:731]: Necessary for the build wheel to pass so we can import cutlass library correctly.

File: cutlass/nvfp4_scaled_mm.cu (hunk 11 above)
Comment (jjsjann123) [reply cutlass/nvfp4_scaled_mm.cu]: I did that in a follow up scaled_mm plumbing #4745

File: tests/python_direct/test_cutlass_nvfp4_quantize.py (hunk 8 above)
Comment (rdspring1) [reply tests/python_direct/test_cutlass_nvfp4_quantize.py]: Not sure what you mean. The kernel checks that the input operands are matrices.

File: cutlass/nvf_cutlass.h (hunk 9 above)
Comment (rdspring1) [reply cutlass/nvf_cutlass.h]: Not required for demo.

File: python/nvfuser_direct/cutlass.py (hunk 10 above)
Comment (rdspring1) [reply python/nvfuser_direct/cutlass.py]: Updated API to do this.

**Comment by rdspring1** [comment]
!test

File: cutlass/nvfp4_quantize.cu (hunk 1 above)
Comment (rdspring1) [reply cutlass/nvfp4_quantize.cu]: Removed the quantize kernels.

File: tests/python_direct/test_cutlass_nvfp4_quantize.py (hunk 8 above)
Comment (crcrpar) [reply tests/python_direct/test_cutlass_nvfp4_quantize.py]: I meant to write 3D instead of the second 2D

File: python/nvfuser_direct/__init__.py [hunk 13]
```diff
@@ -23,6 +23,8 @@
 from . import _C_DIRECT  # noqa: F401,F403
 from ._C_DIRECT import *  # noqa: F401,F403
 
+import torch
```
Comment (jjsjann123) [inline python/nvfuser_direct/__init__.py]: do we need this?

File: python/python_direct/bindings.h [hunk 14]
```diff
@@ -33,4 +33,7 @@ void bindMultiDevice(py::module& nvfuser);
 // Translate a CPP Fusion to a bindings python function
 std::string translateFusion(Fusion* f);
 
+// Add bindings for Cutlass GEMM Operations
+void bindCutlass(py::module& nvfuser);
```
Comment (jjsjann123) [inline python/python_direct/bindings.h:38]: We should have this guarded by `NVFUSER_ENABLE_CUTLASS`.

Do we have a test on building with disabled `NVFUSER_ENABLE_CUTLASS`?

File: tests/python_direct/test_cutlass_nvfp4_gemm.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    x = _f32_to_floatx_unpacked(x.float(), ebits=2, mbits=1)
+    x = pack_uint4(x)
+    x = x.view(torch.float4_e2m1fn_x2)
+    return x
+
+
+def pytorch_nvfp4_quantize(a, a_global_scale):
+    BLOCK_SIZE = 16
+    assert (
+        a.size(-1) % BLOCK_SIZE == 0
```
Comment (jjsjann123) [inline tests/python_direct/test_cutlass_nvfp4_gemm.py:139]: I know I did this as well in my PR :laughing: 

I'm wondering if we should implement padding to lift this requirement.

File: tests/python_direct/test_cutlass_nvfp4_gemm.py [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    ), "inner-most dim must be divisible by block_size"
```
Comment (jjsjann123) [inline tests/python_direct/test_cutlass_nvfp4_gemm.py]: nitpick: let's change the warning to a not-implemented thing instead.

**Review by jjsjann123 — APPROVED** [review]
Note: PR description still mentions `cutlass_nvfp4_quantize`, let's remove that.

File: python/python_direct/bindings.h (hunk 14 above)
Comment (rdspring1) [reply python/python_direct/bindings.h:38]: > Do we have a test on building with disabled NVFUSER_ENABLE_CUTLASS?

I thought the build, A100, and H100 tests would fail if this wasn't correct.
This Github CI build also disables CUTLASS.

File: tests/python_direct/test_cutlass_nvfp4_gemm.py (hunk 15 above)
Comment (rdspring1) [reply tests/python_direct/test_cutlass_nvfp4_gemm.py:139]: I copied from you.

File: python/python_direct/bindings.h (hunk 14 above)
Comment (jjsjann123) [reply python/python_direct/bindings.h:38]: ah, nvm... this is just an declaration.

File: python/utils.py [hunk 17]
```diff
@@ -464,10 +468,10 @@ def on_or_off(flag: bool) -> str:
         f"-DNVFUSER_EXPLICIT_ERROR_CHECK={on_or_off(config.explicit_error_check)}",
         f"-DBUILD_TEST={on_or_off(not config.no_test)}",
         f"-DBUILD_PYTHON={on_or_off(not config.no_python)}",
+        f"-DBUILD_CUTLASS={on_or_off(not config.no_cutlass)}",
         f"-DPython_EXECUTABLE={sys.executable}",
         f"-DBUILD_NVFUSER_BENCHMARK={on_or_off(not config.no_benchmark)}",
         f"-DNVFUSER_DISTRIBUTED={on_or_off(not config.build_without_distributed)}",
-        f"-DUSE_SYSTEM_NVTX={on_or_off(config.build_with_system_nvtx)}",
```
Comment (rdspring1) [inline python/utils.py:470]: I removed this because it isn't used in CMakeLists.txt

**Comment by rdspring1** [comment]
!test
